File: 1057-Campus Bikes.py
from heapq import heappush, heapify
class Solution(object):
    def assignBikes(self, workers, bikes):
        """
        :type workers: List[List[int]]
        :type bikes: List[List[int]]
        :rtype: List[int]
        """
        distance = []
        for i in range(len(workers)):
            distance.append([])
            for j in range(len(bikes)):
                dist = abs(workers[i][0] - bikes[j][0]) + abs(workers[i][1] - bikes[j][1])
                distance[-1].append((dist, i, j))
            distance[-1].sort(reverse = True)
        
        res = [None]*len(workers)
        pq = [x.pop() for x in distance]
        heapify(pq)
        used = set()
        
        while pq:
            _, worker, bike = heappop(pq)
            if bike not in used:
                used.add(bike)
                res[worker] = bike
            else:
                heappush(pq, distance[worker].pop())
        
        return res

File: C++/1057-Campus Bikes.cpp
class Solution {
public:
    int manhattan(vector<int> &a, vector<int> &b){
        return abs(a[0]-b[0]) + abs(a[1]-b[1]);
    }
    
    vector<int> assignBikes(vector<vector<int>>& workers, vector<vector<int>>& bikes) {
        // same shortest distance -> choose smallest worker idx
        // For one worker, many bikes have same distance -> choose smallest bike idx
        vector<pair<int, int>> buckets[2000];
        vector<int> res(workers.size(), -1);
        unordered_set<int> usedWorker;
        unordered_set<int> usedBikes;
        
        for (int i=0; i<workers.size(); i++){
            for (int j=0; j<bikes.size(); j++){
                buckets[manhattan(workers[i], bikes[j])].push_back(pair(i, j));
            }
        }
        
        for (int idx=0; idx<2000; idx++){
            if (buckets[idx].empty()) continue;
            for (auto x:buckets[idx]){
                int i = x.first;
                int j = x.second;
                if (usedWorker.find(i) == usedWorker.end() && usedBikes.find(j) == usedBikes.end()){
                    usedWorker.insert(i);
                    usedBikes.insert(j);
                    res[i] = j;
                    if (usedWorker.size() == workers.size()) return res;
                }
            }
        }
        return res;
    }
};